Implement Quality Gates Resource and Data Type

Bhutania opened the pull request:
# Description: 

This PR implements the quality gates resource and Data types. An example implementation can be found in #76.

reinoudk commented:
@Bhutania awesome stuff already!

I've made some changes to make the tests for the data source work. You can find them on the [terraform-provider-sonarcloud/tree/quality-gates](https://github.com/rewe-digital/terraform-provider-sonarcloud/tree/quality-gates) branch. I hope that you can cherry pick them from there.
 
In general, all the attributes on the new data source are `Computed: true`. `Optional: true` would indicate that these are things that can be provided by the user. (There might be some `Optional: true` in the data source, they should be changed to `Computed: true` as well.)

I've also removed the explicitly defined default values from the schema attributes, they do not need to be specified. Removing them makes the code a bit easier to read imho.

Bhutania commented:
@reinoudk Thank you!

I looked through and implemented the schema changes you are describing. I agree that it makes the code easier to read, they were auto-generated by my IDE and made it easier for me to understand what to specify for each resource, but now that the resources are 'complete' the extra stuff isn't necessary.

I have also changed all the data source attributes to be `Optional: true`. I think I read somewhere that `Computed: true` + `Optional: true` meant that TF doesn't care if the resource is present, but I defer to your experience in the matter 😄 .

reinoudk commented:
> I have also changed all the data source attributes to be `Optional: true`. I think I read somewhere that `Computed: true` + `Optional: true` meant that TF doesn't care if the resource is present, but I defer to your experience in the matter smile .

Hm.. that could make sense yeah. I couldn't find anything conclusive in the docs (maybe I didn't look hard enough though :see_no_evil:). The code comments are not a 100% clear on this either:

```go
// If Required and Optional are both false, Computed must be true, and the
// attribute will be considered "read only" for the practitioner, with
// only the provider able to set its value.
```

Especially the `with only the provider able to set its value` part reads as if both `Required` and `Optional` should be false if you have something that only the provider should be able to set. 

I'll see if I can verify the behavior by testing with a resource that doesn't have any conditions set, i.e:

```tf
resource "sonarcloud_quality_gate" "test" {
	name = "some value"
}
```

reinoudk commented on go.mod:
```diff
@@ -7,7 +7,7 @@ require (
 	github.com/hashicorp/terraform-plugin-go v0.11.0
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.18.0
-	github.com/reinoudk/go-sonarcloud v0.1.1
+	github.com/reinoudk/go-sonarcloud v0.1.2
```
I've released `v0.2.0` of the client, which also shows SonarCloud's error messages when there's a HTTP 4xx/5xx response. This can help a lot with debugging.

```suggestion
	github.com/reinoudk/go-sonarcloud v0.2.0
```

Bhutania commented:
Hi @reinoudk,

I believe that this PR is ready. I will publish this once we resolve https://github.com/reinoudk/go-sonarcloud/pull/5.

I'm excited to see this done, this project taught me a lot about how Terraform works behind the scenes. 😄

Hope you feel well soon!
Arnav

reinoudk commented:
Nice! I've merged https://github.com/reinoudk/go-sonarcloud/pull/5 and `v0.2.1` is available. I'll try and review this tomorrow 🙂.

reinoudk commented on sonarcloud/resource_quality_gate_selection.go:
```diff
@@ -0,0 +1,270 @@
+					tfsdk.RequiresReplace(),
+				},
+			},
+			"project_key": {
```
I'd use the plural `project_keys` here, as it is a set of project keys, not a single project key.

reinoudk commented on examples/resources/sonarcloud_quality_gate_selection/resource.tf:
```diff
@@ -0,0 +1,10 @@
+
+resource "sonarcloud_quality_gate_selection" "example_quality_gate_selection" {
+  gate_id   = data.sonarcloud_quality_gate.awesome_qg.gate_id
+  selection = [for project in data.sonarcloud_projects.all : project.key if project.name == "My Awesome Project"]
```
The resource does not have a `selection` attribute. There is `project_key`, though I'd rename that to `project_keys`, as it is a set of project keys.

reinoudk commented on sonarcloud/data_source_quality_gate.go:
```diff
@@ -0,0 +1,131 @@
+				Computed:    true,
+			},
+			"conditions": {
+				Optional:    true,
```
This can be `Computed: true` as well.

reinoudk commented on sonarcloud/data_source_quality_gate.go:
```diff
@@ -0,0 +1,131 @@
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Could not read the Quality Gate",
+			fmt.Sprintf("The SearchAll request returned an error: %+v", err),
```
Nitpick: this is not a `SearchAll`, but `List` request.

```suggestion
			fmt.Sprintf("The List request returned an error: %+v", err),
```

reinoudk commented on sonarcloud/data_source_quality_gates.go:
```diff
@@ -0,0 +1,142 @@
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Could not read the Quality Gate",
+			fmt.Sprintf("The SearchAll request returned an error: %+v", err),
```
Nitpick: this is not a `SearchAll`, but `List` request.

```suggestion
			fmt.Sprintf("The List request returned an error: %+v", err),
```

reinoudk commented on sonarcloud/resource_quality_gate_selection.go:
```diff
@@ -0,0 +1,270 @@
+
+func (r resourceQualityGateSelectionType) GetSchema(_ context.Context) (tfsdk.Schema, diag.Diagnostics) {
+	return tfsdk.Schema{
+		Description: "This resource selects a quality gate for a project",
```
```suggestion
		Description: "This resource selects a quality gate for one or more projects",
```

reinoudk commented on sonarcloud/resource_quality_gate_selection.go:
```diff
@@ -0,0 +1,270 @@
+			},
+			"gate_id": {
+				Type:        types.StringType,
+				Description: "The ID of the quality gate that is selecting the project(s).",
```
```suggestion
				Description: "The ID of the quality gate that is selected for the project(s).",
```

reinoudk commented on sonarcloud/models.go:
```diff
@@ -50,3 +50,35 @@ type Project struct {
+type Selection struct {
+	ID         types.String `tfsdk:"id"`
+	GateId     types.String `tfsdk:"gate_id"`
+	ProjectKey types.Set    `tfsdk:"project_key"`
```
Use the plural, `ProjectKeys` here.

```suggestion
	ProjectKeys types.Set    `tfsdk:"project_keys"`
```

reinoudk commented on sonarcloud/helpers.go:
```diff
@@ -100,3 +105,52 @@ func findProject(response *projects.SearchResponseAll, key string) (Project, boo
+				ok = true
+				break
+			}
+		}
```
Because we don't check for `ok` after each key, only the last iteration will determine the value of `ok`. What can happen is that the second-to-last project key is not present in the result, but because the last project key **is**, `ok` will be set to `true`, even though we are missing a project selection.

To fix this, we should check for `ok` being `false`, and break the outer loop if it is indeed `false`.

```suggestion
		}
		if !ok {
		  break
		}
```

reinoudk reviewed (changes requested):
There's one bug that caused the acceptance test to fail, but other than it was looking pretty green 🟢 🚀 . 

My other points are mainly concerning style, just some small stuff 🙂 .

Bhutania commented:
I believe that I have implemented the changes you requested. I agree with all of them 😄. 

I have double checked that the changes you have requested are present, but if you want to go through them again just to verify feel free! Looking forward to a (hopefully 😅 ) bug-free release! 🟢  🚀

reinoudk reviewed (approved):
Awesome stuff, and thanks for taking the time and effort to implement this @Bhutania :rocket: !
I'll merge this and then prepare the release. I've had a small hiccup when trying to generate the docs for the new resources, but that should be fixable. 

I'll let you know when the new release is ready!
